requests/models.py

```diff
@@ -365,11 +365,19 @@ def prepare_url(self, url, params):
         if not host:
             raise InvalidURL("Invalid URL %r: No host supplied" % url)
 
-        # Only want to apply IDNA to the hostname
-        try:
-            host = idna.encode(host, uts46=True).decode('utf-8')
-        except (UnicodeError, idna.IDNAError):
-            raise InvalidURL('URL has an invalid label.')
+        # In general, we want to IDNA encode everything, as that allows users
+        # to automatically get correct behaviour in the majority of cases.
+        # However, we’re quite strict about IDNA encoding, so if a user has
+        # already IDNA encoded their domain name, we shouldn’t bother validating
+        # their work. We determine something *looks* IDNA encoded by checking
+        # for the IDNA-encoding tag (xn--) at the start of the hostname, and
+        # confirming that the hostname is entirely ASCII characters.
+        if not host.startswith('xn--') or not is_ascii(host):
```

Comment:
So `unicode` won't work here because it's not defined for Python 3 in compat. We bind `str` to `unicode` for Python 2. However, strings that are declared like `'my string'`, as I noted above, are not `unicode` but will still evaluate to `True` in `isinstance(string, str)`.

I can implement the standalone try/except block, I just wanted to make sure the subtle failure points were understood before making the changes. Either, bytestrings will inconsistently evaluate between versions, or anything not expressly defined `u'string'` or `b'string'.encode('utf-8')` will.